src/output/plugins/MacOSOutputPlugin.cxx

```diff
@@ -0,0 +1,523 @@
+	}
+	pcm_export.Construct();
+}
+
+void
+MacOSOutput::Disable() noexcept {
+	device.Close();
+	pcm_export.Destruct();
+}
+
+void
+MacOSOutput::Close() noexcept {
+	try {
+		device.RemoveIOProc();
+		if(hog_device)
+			// Release hog mode
+			device.SetHogStatus(false);
+	}
+	catch (...) {
+		// Make sure this is really noexcept method.
+		FormatDebug(macos_output_domain,"Ignoring exception on close of output.");
+	}
+	pcm_export->Reset();
+	DeallocateABL(out_buffer);
+	out_buffer = nullptr;
+	AudioConverterDispose(ca_converter);
+	ca_converter = nullptr;
+	delete ring_buffer;
+}
+
+void
+MacOSOutput::SetupConverter() {
+	/** Setup the audio converter in the following cases:
+	 *	1. Integer mode is not used and float conversion is needed
+	 *	2. Number of channels for MPD format and device format differs
+	 *	3. Channel map was specified and therefore mapping / re-ordering required
+	 *	4. Usage of planar audio device (de-interleaving required)
+	 */
+	OSStatus err = noErr;
+	if(!channel_map.empty()) {
+		if(out_format.mChannelsPerFrame > channel_map.size())
+			throw FormatRuntimeError("Channel map contains only %l channels, output device requires %d channels.", channel_map.size(), out_format.mChannelsPerFrame);
+		err = AudioConverterNew(&in_format, &out_format, &ca_converter);
+		/** Pass directly the array indicating with the size parameter
+		 *	the number of channels to be read from the channel_map.
+		 */
+		err = AudioConverterSetProperty(ca_converter, kAudioConverterChannelMap, out_format.mChannelsPerFrame * sizeof(SInt32), channel_map.data());
+
+	}
+	else if(!(out_format.mFormatFlags & kAudioFormatFlagIsNonMixable) || device.IsPlanar() || in_format.mChannelsPerFrame != out_format.mChannelsPerFrame) {
+		// Integer mode not active, planar device or channel conversion needed
+		err = AudioConverterNew(&in_format, &out_format, &ca_converter);
+	}
+	
+	if(err != noErr)
+		throw std::runtime_error("Failed to setup AudioConverter for MacOS output.");
+	
+	/** Allocate the buffer used for output conversion.
+	 *	Since the CoreAudio HAL asks as max for each
+	 *	callback for the number of frames that are set
+	 *	as the (variable) device buffer size, exactly
+	 *	this amount gets allocated here.
+	 */
+	try {
+		if(ca_converter != nullptr)
+			out_buffer = AllocateABL(in_format, device.GetBufferSize());
+	}
+	catch (...) {
+		AudioConverterDispose(ca_converter);
+		ca_converter = nullptr;
+		throw;
+	}
+}
+
+void
+MacOSOutput::Setup(AudioFormat &audio_format) {
+	if (device.SetFormat(audio_format, integer_mode)) {
+		/** Report back the actual physical device format
+		 *	to make sure that MPDs output engine sends
+		 *	the physical format. This will get transformed
+		 *	to the virtual format in final conversion
+		 *	directly in the render callback (to float
+		 *	samples in case integer_mode is not
+		 *	configured or not supported).
+		 */
+		
+		memset(&in_format, 0, sizeof(in_format));
+		memset(&out_format, 0, sizeof(out_format));
```

Comment:
why memset is needed? in/out_format will be completely copied from device.GetPhys/IOFormat

Reply:
Thanks, makes sense to delete the memset calls indeed!